main/src/ca/uwaterloo/flix/language/phase/TailrecCheck.scala

```diff
@@ -0,0 +1,170 @@
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.CompilationError
+import ca.uwaterloo.flix.language.ast.Ast.Annotation
+import ca.uwaterloo.flix.language.ast.TypedAst
+import ca.uwaterloo.flix.language.ast.TypedAst.Expression
+import ca.uwaterloo.flix.language.errors.NonTailRecursiveCallError
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess}
+
```

Comment:
So, while this approach works, I wonder if it is a good idea to have a separate phase just for this. Or at least if it should not be combined with the Tailrec phase later in the compiler?